### PR TITLE
[AIR] Improve `Normalizer` docstring

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

These changes are needed to make `Normalizer` easier to understand.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/preprocessors/normalizer.py [hunk 1]
```diff
@@ -7,16 +7,69 @@
...
+        s' = \frac{s}{\lVert s \rVert_p}
+
+    where :math:`s` is the sample, :math:`s'` is the transformed sample, and :math:`p`
+    is the norm type.
+
+    The following norms are supporting:
+
+        * :math:`L^1`: Sum of the absolute values.
+        * :math:`L^2`: Square root of the sum of the squared values.
+        * :math:`L^\infty`: Maximum value.
```
Comment (bveeramani) [inline python/ray/data/preprocessors/normalizer.py:27]: I want to explain why you'd use one norm over another, but I don't know. Anyone know the answer?

File: python/ray/data/preprocessors/normalizer.py [hunk 2]
```diff
@@ -7,16 +7,69 @@
...
 
     Args:
-        columns: The columns that in combination define the record to normalize.
-        norm: "l1", "l2", or "max". Defaults to "l2"
+        columns: The columns to include in samples.
```
Comment (bveeramani) [inline python/ray/data/preprocessors/normalizer.py:29]: Wasn't really sure how to explain this clearly. It's confusing how the operation is applied to rows but we specify columns

File: python/ray/data/preprocessors/normalizer.py [hunk 3]
```diff
@@ -7,16 +7,69 @@
...
+        >>>
+        >>> df = pd.DataFrame({"X1": [1, 1], "X2": [1, 0], "X3": [0, 1]})
+        >>> ds = ray.data.from_pandas(df)  # doctest: +SKIP
+        >>> ds.to_pandas()  # doctest: +SKIP
+           X1  X2  X3
+        0   1   1   0
+        1   1   0   1
+
+        The :math:`L^2`-norm of the first sample is :math:`\sqrt{2}`, and the
+        :math:`L^2`-norm of the second sample is :math:`1`.
```
Comment (bveeramani) [inline python/ray/data/preprocessors/normalizer.py:49]: Don't know how useful these descriptions are. Could use feedback on how to make the example descriptions more useful

File: python/ray/data/preprocessors/normalizer.py [hunk 4]
```diff
@@ -7,16 +7,69 @@
...
+    The general formula is given by
+
+    .. math::
+
+        s' = \frac{s}{\lVert s \rVert_p}
+
+    where :math:`s` is the sample, :math:`s'` is the transformed sample, and :math:`p`
+    is the norm type.
+
+    The following norms are supporting:
```
Comment (matthewdeng) [inline python/ray/data/preprocessors/normalizer.py:22]:
```suggestion
    The following norm types are supported:
```

File: python/ray/data/preprocessors/normalizer.py [hunk 5]
```diff
@@ -7,16 +7,69 @@
...
-        * max: Maximum value.
+    This preprocessor works by dividing each sample (i.e., row) by the sample's norm.
+    The general formula is given by
+
+    .. math::
+
+        s' = \frac{s}{\lVert s \rVert_p}
+
+    where :math:`s` is the sample, :math:`s'` is the transformed sample, and :math:`p`
+    is the norm type.
```
Comment (matthewdeng) [inline python/ray/data/preprocessors/normalizer.py:20]: Mention that `||s||` is the norm?

File: python/ray/data/preprocessors/normalizer.py [hunk 6]
```diff
@@ -7,16 +7,69 @@
...
+
+    The following norms are supporting:
+
+        * :math:`L^1`: Sum of the absolute values.
+        * :math:`L^2`: Square root of the sum of the squared values.
+        * :math:`L^\infty`: Maximum value.
```
Comment (matthewdeng) [inline python/ray/data/preprocessors/normalizer.py:26]: Would be helpful to map this directly to the user input, something like:

```suggestion
        * `"l1"` (:math:`L^1`): Sum of the absolute values.
        * `"l2"` (:math:`L^2`): Square root of the sum of the squared values.
        * `"max"` (:math:`L^\infty`): Maximum value.
```

Comment (matthewdeng) [reply python/ray/data/preprocessors/normalizer.py:26]: In particular I don't think it's obvious that `L^infinity == max`

File: python/ray/data/preprocessors/normalizer.py (hunk 2 above)
Comment (matthewdeng) [reply python/ray/data/preprocessors/normalizer.py:29]: Yeah.... maybe something like "For each row, these the values of these columns will together constitute the sample to scale".

File: python/ray/data/preprocessors/normalizer.py [hunk 7]
```diff
@@ -7,16 +7,69 @@
...
+        :math:`L^2`-norm of the second sample is :math:`1`.
+
+        >>> preprocessor = Normalizer(columns=["X1", "X2"])  # doctest: +SKIP
+        >>> preprocessor.fit_transform(ds).to_pandas()  # doctest: +SKIP
+                 X1        X2  X3
+        0  0.707107  0.707107   0
+        1  1.000000  0.000000   0
+
+        The :math:`L^1`-norm of the first sample is :math:`2`, and the
+        :math:`L^2`-norm of the second sample is :math:`1`.
```
Comment (matthewdeng) [inline python/ray/data/preprocessors/normalizer.py:58]:
```suggestion
        :math:`L^1`-norm of the second sample is :math:`1`.
```

File: python/ray/data/preprocessors/normalizer.py (hunk 3 above)
Comment (matthewdeng) [reply python/ray/data/preprocessors/normalizer.py:49]: Maybe add it directly as a comment in the code block?

File: python/ray/data/preprocessors/normalizer.py [hunk 8]
```diff
@@ -7,16 +7,69 @@
...
+
+        >>> preprocessor = Normalizer(columns=["X1", "X2"], norm="l1")  # doctest: +SKIP
+        >>> preprocessor.fit_transform(ds).to_pandas()  # doctest: +SKIP
+            X1   X2  X3
+        0  0.5  0.5   0
+        1  1.0  0.0   1
+
+        The :math:`L^\infty`-norm of the both samples is :math:`1`,
```
Comment (matthewdeng) [inline python/ray/data/preprocessors/normalizer.py:66]:
```suggestion
        The :math:`L^\infty`-norm of the both samples is :math:`1`.
```

File: python/ray/data/preprocessors/normalizer.py (hunk 4 above)
Comment (bveeramani) [reply python/ray/data/preprocessors/normalizer.py:22]: Fixed. Don't think the word "types" is necessary.`l1` are `l2` are definitionally norms.

File: python/ray/data/preprocessors/normalizer.py (hunk 5 above)
Comment (bveeramani) [reply python/ray/data/preprocessors/normalizer.py:20]: Fixed

File: python/ray/data/preprocessors/normalizer.py (hunk 2 above)
Comment (bveeramani) [reply python/ray/data/preprocessors/normalizer.py:29]: Fixed

**Comment by bveeramani** [comment]
Closing in favor of #27215
